### PR TITLE
core_tests: more tests and some refactoring

### PR DESCRIPTION
**Pull request by msgmaxim**
@Doy-lee please review

### CODE REVIEW

File: src/cryptonote_core/service_node_list.cpp [hunk 1]
```diff
@@ -729,7 +729,7 @@ namespace service_nodes
     return  x / (secureMax / n);
   }
 
-  static void loki_shuffle(std::vector<size_t>& a, uint64_t seed)
+  void loki_shuffle(std::vector<size_t>& a, uint64_t seed)
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply src/cryptonote_core/service_node_list.cpp]: I don't have access to a service node list (on generation), so unless it becomes a namespace function, we wouldn't be able to use it there.

[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_core/service_node_list.cpp]: That sounds reasonable, static member function or just in the namespace sounds fine.

File: tests/core_tests/service_nodes.cpp [hunk 2]
```diff
@@ -36,7 +36,190 @@ using namespace std;
...
+      events_.push_back(blk);
+    }
+
+    void create_block_sn(const std::vector<cryptonote::transaction>& txs = {})
+    {
+
+      const auto winner = sn_owners_.at(next_winner_idx);
+
+      next_winner_idx++;
+      if (next_winner_idx >= sn_owners_.size()) {
```
Comment (Doy-lee) [inline tests/core_tests/service_nodes.cpp]: Also, getting signed vs unsigned comparison warning here.

**Comment by msgmaxim** [comment]
@Doy-lee The last commit does some refactoring to the linear chain generator. Importantly, it correctly chooses the service node winner allowing more complex scenarios (e.g. when nodes deregister at different times).

**Comment by msgmaxim** [comment]
New commits add two more core tests for service nodes.

File: tests/core_tests/service_nodes.cpp [hunk 3]
```diff
@@ -36,15 +36,84 @@ using namespace std;
...
+}
 
 struct last_reward_point {
   uint64_t height;
   uint64_t priority;
 };
 
-bool operator < (const last_reward_point& lhs, const last_reward_point& rhs) {
-  return (lhs.height < rhs.height) || (lhs.priority < rhs.priority);
+bool operator<(const last_reward_point& lhs, const last_reward_point& rhs) {
```
Comment (msgmaxim) [inline tests/core_tests/service_nodes.cpp]: This was previously incorrect and worked only by chance.

File: tests/core_tests/service_nodes.cpp [hunk 4]
```diff
@@ -36,53 +36,531 @@ using namespace std;
...
+
+    cryptonote::transaction create_registration_tx();
+
+    const cryptonote::account_base& first_miner() const { return first_miner_; }
+
+    /// Note: should be carefull with returing a reference to vector elements
+    const cryptonote::block& chain_head() const { return blocks_.back(); }
+
+    /// get a copy of the service node list
+    const sn_list get_sn_list() const { return sn_list_; }
```
[the remark this answers is not in the document]
Comment (msgmaxim) [reply tests/core_tests/service_nodes.cpp]: There used to be a guideline advocating returning const T (from *Effective C++*), but it appears to be obsolete now. Happy to remove the `const`.

**Comment by jcktm** [comment]
Please rebase

**Pull request merged**